src/webgpu/api/validation/vertex_state.spec.ts

```diff
@@ -488,183 +459,202 @@ g.test('vertex_shader_input_location_in_vertex_state')
+    - Test for various vertex buffer indices
+    - Test for various amounts of attributes in that vertex buffer`
+  )
+  .cases(
+    params()
+      .combine(poptions('format', kVertexFormats))
+      .combine(
+        poptions('arrayStride', [
+          0,
+          256,
+          kMaxVertexBufferArrayStride - 4,
+          kMaxVertexBufferArrayStride,
+        ])
+      )
+      .expand(p => {
+        // Compute a bunch of test offsets to test.
+        const { bytesPerComponent, componentCount } = kVertexFormatInfo[p.format];
+        const formatSize = bytesPerComponent * componentCount;
+        const offsetsToTest = [0, bytesPerComponent];
+
+        // Avoid adding offset based on the arrayStride if it would make duplicated tests at it
+        // makes the harness assert.
+        if (p.arrayStride !== 0) {
+          offsetsToTest.push(p.arrayStride - formatSize);
+          offsetsToTest.push(p.arrayStride - formatSize + bytesPerComponent);
+        } else {
+          offsetsToTest.push(kMaxVertexBufferArrayStride - formatSize);
+          offsetsToTest.push(kMaxVertexBufferArrayStride - formatSize + bytesPerComponent);
+        }
+
+        return poptions('offset', offsetsToTest);
+      })
+  )
+  .subcases(() =>
+    params()
+      .combine(poptions('vertexBufferIndex', [0, 1, kMaxVertexBuffers - 1]))
+      .combine(poptions('extraAttributeCount', [0, 1, kMaxVertexAttributes - 1]))
+      .combine(pbool('testAttributeAtStart'))
+  )
+  .fn(t => {
+    const {
+      format,
+      arrayStride,
+      offset,
+      vertexBufferIndex,
+      extraAttributeCount,
+      testAttributeAtStart,
+    } = t.params;
+
+    const attributes: GPUVertexAttributeDescriptor[] = [];
+    addTestAttributes(attributes, {
+      testAttribute: { format, offset, shaderLocation: 0 },
+      testAttributeAtStart,
+      extraAttributeCount,
+      extraAttributeSkippedLocations: [0],
     });
-  }
-});
 
-g.test('check_attribute_offset_overflow').fn(async t => {
-  const vertexState: GPUVertexStateDescriptor = {
-    vertexBuffers: [
-      {
-        arrayStride: 0,
-        attributes: [{ format: 'float32', offset: Number.MAX_SAFE_INTEGER, shaderLocation: 0 }],
-      },
-    ],
-  };
-  const descriptor = t.getDescriptor(vertexState, VERTEX_SHADER_CODE_WITH_NO_INPUT);
+    const vertexBuffers = [];
+    vertexBuffers[vertexBufferIndex] = { arrayStride, attributes };
 
-  t.expectValidationError(() => {
-    t.device.createRenderPipeline(descriptor);
+    const formatInfo = kVertexFormatInfo[format];
+    const formatSize = formatInfo.bytesPerComponent * formatInfo.componentCount;
+    const limit = arrayStride === 0 ? kMaxVertexBufferArrayStride : arrayStride;
+
+    const success = offset + formatSize <= limit;
+    t.testVertexState(success, { vertexBuffers });
+  });
+
+g.test('many_attributes_overlapping')
+  .desc(`Test that it is valid to have many vertex attributes overlap`)
```

Comment:
nit: Maybe test with different formats (like rotate the `format` between the attributes)?

Or maybe this should be an operation test later that actually tests the load behavior? (or both validation and operation test)

Reply:
There should be an operation test that does complicated cases like this, but I still added the format rotation.